### PR TITLE
fix(repl): render the unjsonable things described in #2278

### LINKED ISSUE #2278: REPL support for Symbol, undefined, NaN, and infinities
## What is the Problem Being Solved?

in #2261, the REPL's `stringify()` was fixed to support BigInts, and @erights observed that it would fail similarly for symbols, `undefined`, `NaN`, `Infinity`, and `-Infinity` "all of which are primitive values that don't encode in JSON."

## Description of the Design

Replace the special case branch with a replacer in the `JSON.stringify()` call that handles all these primitives.

## Security Considerations

The REPL won't be able to report values containing these kinds of primitives. This results in an error, so it won't go unnoticed.

## Test Plan

Added a failing test for NaN in `test-repl.js`

### PR DESCRIPTION
**Pull request by michaelfig**
Closes #2278 

This uses a map to keep track of the names for unjsonable things, with some special handling for symbols.

### CODE REVIEW

File: packages/vats/src/repl.js [hunk 1]
```diff
@@ -17,6 +32,13 @@ export function stringify(
     if (typeof value === 'bigint') {
       return `${value}n`;
     }
+    const rawString = UNJSONABLES.get(value);
+    if (rawString) {
+      return rawString;
+    }
+    if (typeof value === 'symbol') {
+      return `Symbol(?)`;
```
Comment (erights) [inline packages/vats/src/repl.js]: Symbols carry a description string. Registered symbols have their registration string (which is probably also their description string). So you can give a lot more info that `?`. How about `String(value)` ?

**Review by erights — COMMENTED** [review]
I am concerned about the proliferation of these. I see you're still using `JSON.stringify` with a replacer, and so living with those limitations. Given that, what is the advantage of this one over `bestEffortStringify` at

https://github.com/endojs/endo/blob/f7dcf050ef0fd839ab6500b938791740c9bc702b/packages/ses/src/error/stringify-utils.js#L22

which is already available indirectly as 
```js
`${q(value)}`
```
Whatever those advantages are, can they be consolidated into `bestEffortStringify` to minimize this proliferation?

**Comment by michaelfig** [comment]
> I am concerned about the proliferation of these. I see you're still using `JSON.stringify` with a replacer, and so living with those limitations.

I'm not using `JSON.stringify` with a replacer.  I'm specifically generating strings that are unambiguous but definitely not valid JSON when JSON cannot express what I've found.

> Given that, what is the advantage of this one over `bestEffortStringify` at
> 
> https://github.com/endojs/endo/blob/f7dcf050ef0fd839ab6500b938791740c9bc702b/packages/ses/src/error/stringify-utils.js#L22

`bestEffortStringify` always produces valid JSON, even for things that are not jsonable.

> Whatever those advantages are, can they be consolidated into `bestEffortStringify` to minimize this proliferation?

I'd love to consolidate.  Do you have any ideas how to accomplish that?

**Comment by erights** [comment]
Sorry, I saw
```js
return JSON.stringify(value, null, spaces);
```
forgot the order of arguments, assumed that `spaces` was a replacer, and then jumped to a lot of wrong conclusions. I'll start again.

File: packages/vats/src/repl.js [hunk 2]
```diff
@@ -5,6 +5,21 @@ import { getInterfaceOf, Remotable, Far } from '@agoric/marshal';
...
+
+for (const [name, { value }] of Object.entries(
+  Object.getOwnPropertyDescriptors(Symbol),
+)) {
+  if (typeof value === 'symbol' && typeof name === 'string') {
+    UNJSONABLES.set(value, `Symbol(Symbol.${name})`);
+  }
+}
+
 // A REPL-specific JSON stringify.
```
Comment (erights) [inline packages/vats/src/repl.js]: This comment contributed to my misunderstanding that this is an *adjective* JSON stringify, for some adjective.

File: packages/vats/src/repl.js (hunk 1 above)
Comment (michaelfig) [reply packages/vats/src/repl.js]: TIL:
```js
String(someSymbol) !== `${someSymbol}`
```

Thanks.  I've updated that now.

File: packages/vats/src/repl.js (hunk 2 above)
Comment (michaelfig) [reply packages/vats/src/repl.js]: I'll clear that up.

Comment (michaelfig) [reply packages/vats/src/repl.js]: Renamed `stringify` to `dump(obj, spaces = undefined)`.

**Review by erights — APPROVED** [review]
One bug: The missing `inProgress.delete`. Everything else is minor.

LGTM.

If you don't mind, I'm curious what your output looks like on the challenges at https://github.com/endojs/endo/blob/ccde56bc5d476633cc816ed594867e9cc8bac5c8/packages/ses/test/error/test-assert-log.js#L414

That would help make these more similar, or at least understand better where they differ.

File: packages/vats/src/repl.js [hunk 3]
```diff
@@ -12,14 +12,13 @@ const UNJSONABLES = new Map([
...
-  depth = 0,
-) {
+// A REPL-specific data dump-to-string.  This specifically is *not* JSON, but its
+// output is unambiguous (even though it cannot be round-tripped).
+
+// eslint-disable-next-line no-use-before-define
+export const dump = (value, spaces = undefined) => dump0(value, spaces);
+
+function dump0(value, spaces, inProgress = new WeakSet(), depth = 0) {
   if (Object(value) !== value) {
```
[the remark this answers is not in the document]
Comment (michaelfig) [reply packages/vats/src/repl.js:22]: I'll save it for a rainy day.

File: packages/vats/src/repl.js [hunk 4]
```diff
@@ -42,7 +55,7 @@ export function stringify(
 
   let ret = '';
   const spcs = spaces === undefined ? '' : ' '.repeat(spaces);
-  if (gio && gio(value) !== undefined) {
+  if (getInterfaceOf(value) !== undefined) {
     ret += `${value}`;
```
[the remark this answers is not in the document]
Comment (michaelfig) [reply packages/vats/src/repl.js]: I want to dump the contents of objects that have a string tag, so I kept `[Promise]` separately-handled.

**Comment by michaelfig** [comment]
> If you don't mind, I'm curious what your output looks like on the challenges at https://github.com/endojs/endo/blob/ccde56bc5d476633cc816ed594867e9cc8bac5c8/packages/ses/test/error/test-assert-log.js#L414

After some tweaking, I got it closer to what Node reports on the console:

```js
  t.is(
    dump(challenges),
    '[[Promise],[Function foo],"[hilbert]",undefined,"undefined",[URIError: wut?],[33n,Symbol(foo),Symbol(bar),Symbol(Symbol.asyncIterator)],{"NaN":NaN,"Infinity":Infinity,"neg":-Infinity},18014398509481984,{"superTagged":{[Symbol(Symbol.toStringTag)]:"Tagged"},"subTagged":[Object Tagged]{},"subTaggedNonEmpty":[Object Tagged]{"foo":"x"}}]',
  );
  t.is(
    dump(challenges, '  '),
    `\
[
  [Promise],
  [Function foo],
  "[hilbert]",
  undefined,
  "undefined",
  [URIError: wut?],
  [
    33n,
    Symbol(foo),
    Symbol(bar),
    Symbol(Symbol.asyncIterator)
  ],
  {
    "NaN": NaN,
    "Infinity": Infinity,
    "neg": -Infinity
  },
  18014398509481984,
  {
    "superTagged": {
      [Symbol(Symbol.toStringTag)]: "Tagged"
    },
    "subTagged": [Object Tagged] {},
    "subTaggedNonEmpty": [Object Tagged] {
      "foo": "x"
    }
  }
]`,
  );
```

**Pull request merged**